### PR TITLE
stylo: Support linear-gradients as background-image

### PR DESCRIPTION
**Pull request by emilio**
This PR supports setting the background-image property to a linear-gradient in Geckolib.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes do not require tests because these are geckolib-only changes.

r? @mbrubeck 
cc @heycam @bholley

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11456)

### CODE REVIEW

**Comment by emilio** [comment]
Gecko bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1276085

File: ports/geckolib/properties.mako.rs [hunk 1]
```diff
@@ -700,10 +705,134 @@ fn static_assert() {
...
+                                                       &mut (*gecko_gradient).mAngle.mValue);
+                        }
+                    }
+
+                    let mut coord: nsStyleCoord = unsafe { uninitialized() };
+                    for (index, stop) in gradient.stops.iter().enumerate() {
+                        // NB: stops are guaranteed to be none in the gecko side by
+                        // default.
+                        stop.position.to_gecko_style_coord(&mut coord.mUnit,
+                                                           &mut coord.mValue);
```
Comment (mbrubeck) [inline ports/geckolib/properties.mako.rs]: `coord.set(stop.position)`

File: ports/geckolib/properties.mako.rs [hunk 2]
```diff
@@ -700,10 +705,134 @@ fn static_assert() {
...
+
+                    if gecko_gradient.is_null() {
+                        warn!("stylo: Couldn't allocate gradient");
+                        return;
+                    }
+
+                    if let AngleOrCorner::Angle(angle) = gradient.angle_or_corner {
+                        unsafe {
+                            angle.to_gecko_style_coord(&mut (*gecko_gradient).mAngle.mUnit,
+                                                       &mut (*gecko_gradient).mAngle.mValue);
```
Comment (mbrubeck) [inline ports/geckolib/properties.mako.rs]: You can write this as `gecko_gradient.mAngle.set(angle);`.

**Comment by mbrubeck** [comment]
r=mbrubeck with some minor cleanup suggestions (above).

**Comment by emilio** [comment]
@mbrubeck: Thanks for the fast review! Forgot about those methods :)

Addressed the comments, maybe worth waiting for @heycam's review in Gecko to see if any other change is required.

**Comment by emilio** [comment]
Gecko changes got r+'d, so I'm going to land this (per the `r=mbrubeck` above) and make the nsTArray bindings work in a separate PR today.

@bors-servo: r=mbrubeck

**Comment by emilio** [comment]
@bors-servo:  retry

 * Possibly new intermittent?

**Comment by emilio** [comment]
Added the nsTArray bindings, had to add a patch from @bholley that fixes overflow-wrap with the new bindings. cc @heycam about that.

Note that this won't build until #11445 lands, since it uses `copy_from()`.

re-r? @mbrubeck

**Comment by mbrubeck** [comment]
-S-blocked-on-external -S-awaiting-review +S-awaiting-answer 

Looks good to me, with one suggested change. r=mbrubeck with or without the change.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.
</details>


---

Reviewed 5 of 5 files at r1, 1 of 1 files at r2, 6 of 6 files at r3, 2 of 9 files at r6, 4 of 6 files at r7, 1 of 1 files at r8, 1 of 2 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[ports/geckolib/gecko_bindings/sugar/ns_t_array.rs, line 9 \[r3\]](https://reviewable.io:443/reviews/servo/servo/11456#-KJGom1mh1FZWq75fwrY:-KJGom1n2fuDxlHXT1L6:1164355462) ([raw file](https://github.com/servo/servo/blob/0d9455ec1399299ae24e2914b98da71ec1751a2d/ports/geckolib/gecko_bindings/sugar/ns_t_array.rs#L9)):*
> ```Rust
> use structs::{nsTArray, nsTArrayHeader};
> 
> impl<T> Index<u32> for nsTArray<T> {
> ```

Would it make sense to instead impl Deref/DerefMut with `Target=[T]` for `nsTArray<T>`? Then we would get `len`, `Index`, `IndexMut` and other slice methods for free.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/11456#-:-KJGpiViLrGbhd13h_cg:-1357008967)*

**Comment by emilio** [comment]
Yep, that's a great idea, I don't know why I didn't thought about it :)

@bors-servo: r=mbrubeck

**Pull request merged**